src/node_crypto.cc

```diff
@@ -3281,21 +3376,36 @@ bool CipherBase::Final(unsigned char** out, int *out_len) {
   if (ctx_ == nullptr)
     return false;
 
+  const int mode = EVP_CIPHER_CTX_mode(ctx_);
+
   *out = Malloc<unsigned char>(
       static_cast<size_t>(EVP_CIPHER_CTX_block_size(ctx_)));
-  int r = EVP_CipherFinal_ex(ctx_, *out, out_len);
 
-  if (r == 1 && kind_ == kCipher && IsAuthenticatedMode()) {
-    auth_tag_len_ = sizeof(auth_tag_);
-    r = EVP_CIPHER_CTX_ctrl(ctx_, EVP_CTRL_GCM_GET_TAG, auth_tag_len_,
-                            reinterpret_cast<unsigned char*>(auth_tag_));
-    CHECK_EQ(r, 1);
+  // In CCM mode, final() only checks whether authentication failed in update().
+  // EVP_CipherFinal_ex must not be called and will fail.
+  bool ok;
+  if (kind_ == kDecipher && mode == EVP_CIPH_CCM_MODE) {
+    ok = !pending_auth_failed_;
+  } else {
+    ok = EVP_CipherFinal_ex(ctx_, *out, out_len) == 1;
+
+    if (ok && kind_ == kCipher && IsAuthenticatedMode()) {
+      // For GCM, the tag length is static (16 bytes), while the CCM tag length
+      // must be specified in advance.
+      if (mode == EVP_CIPH_GCM_MODE)
+        auth_tag_len_ = sizeof(auth_tag_);
+      // TOOD(tniessen) Use EVP_CTRL_AEAP_GET_TAG in OpenSSL 1.1.0
```

Comment:
TODO needs to be done, we have to support 1.1.0

Reply:
See https://github.com/nodejs/node/pull/18138#discussion_r166016423.